libgadget/fof.c

```diff
@@ -382,40 +373,96 @@ void fof_label_primary(void)
+ * and merges 2 into h1 by setting Head[h2] = h1. It does this if and only if h1 < h2 (to avoid cycles)
+ * and it does it under a lock for h2, so that h2 cannot be merged elsewhere. h1 is not locked, to avoid deadlocking.
+ * It can thus cease to be the head of the tree, but this is fine, because it will never be merged into h2.
+ * If h1 > h2, the code swaps them and tries again.
+ * returns: 1 if a merge happened, 0 if it did not. No locks are held after return.*/
+static int
+try_merge_trees(int * h1, int * h2, MyIDType * subminID, MyIDType * subminIDTask, int * Head)
+{
+    /* We are about to change the HEAD of h2, so lock it.
+     * We need to make sure that h2 is really the head of the subtree,
+     * or not all particles will be linked.
+     * If this is not the case, unlock the particle, walk the tree again,
+     * and loop until we can get a lock.*/
+    while(1) {
+        /*Stop if we encountered h1 along the head of this tree: means this is already merged.*/
+        if(*h2 == -1)
+            return 0;
+        lock_particle(*h2);
+        if(Head[*h2] != *h2) {
+            unlock_particle(*h2);
+            *h2 = HEADa(*h1, *h2, Head);
+        }
+        else
+            break;
+    }
+    /*Get the current subtree properties: this is fine because we are under the lock.*/
+    *subminID = HaloLabel[*h2].MinID;
+    *subminIDTask = HaloLabel[*h2].MinIDTask;
+
+    /*Merge the subtree*/
+    if(*h2 > *h1) {
+        #pragma omp atomic write
+        Head[*h2] = *h1;
+        unlock_particle(*h2);
+        return 1;
+    }
+    else {
+        /*If h1 > h2, swap the particles and try again*/
+        unlock_particle(*h2);
+        int t = *h2;
+        *h2 = *h1;
+        *h1 = t;
+        return try_merge_trees(h1, h2, subminID, subminIDTask, Head);
+    }
+}
+
 static void
 fofp_merge(int target, int other, TreeWalk * tw)
 {
-    /* this will lock h1 */
-    int h1 = HEADl(-1, target, tw);
-    /* stop looking if we find h1 along the path (because it is already owned by us) */
-    int h2 = HEADl(h1, other, tw);
-
-    if(h2 == -1) {
-        /* h1 is along the path of h2, already merged.  **/
-        /* h1 must be the root of other and target both */
-        //printf("unlocking %d by %d in merge\n", h1, omp_get_thread_num());
-        update_root(target, h1, tw);
-        update_root(other, h1, tw);
-        unlock_particle(h1);
-        return;
+    int * Head = FOF_PRIMARY_GET_PRIV(tw)->Head;
+    if(other < target)
+    {
+        int t = target;
+        target = other;
+        other = t;
     }
+    /* Get current HEAD of target. h1 is not locked, so by the time we use it,
+     * it may no longer really be the HEAD, but we could also link the
+     * tree by setting HEAD(h2) = target, so it doesn't matter.*/
+    int h1 = HEADa(-1, target, Head);
+    int h2 = HEADa(h1, other, Head);
 
-    /* h2 as a sub-tree of h1 */
-    FOF_PRIMARY_GET_PRIV(tw)->Head[h2] = h1;
```

Comment:
Not sure. We need to enforce that there is an ordering somehow, or we can get h2 added below h1 by another thread while we are waiting for the lock.

Reply:
Here we only hold at most one lock at any time. 

Could you be more elaborate `waiting for the lock`? which lock?
h2 is locked and unlocked during the last three lines.

Reply:
Yes I just ran into this. The ordering would solve this. I am leaning towards your solution is the only solution. But let me think about this a bit more.

Reply:
Yes. I bet splays can be added. with easy.

My main reservation over your algorithm is that it feels trickier (harder to elaborate) than the existing one. This viewpoint must have been biased, as I am more familiar with the old algoirthm, but it enhanced my worry that there are corners that's not yet caught.

I think a retry algorithm is needed if we convert a double lock to two single locks. (not sure if there is a theorem, but it does smell like the right complexity mentioned in that paper). 

In this problem, the retry lock would be implemented like this: as we try to lock the second particle, we if we encounter any particle that is already locked, then we lock at the index:
- If the index of the newly found is less than the locked one we already have, wait on the lock.
- If not, unlock the first locked particle, and retry, optionally sleep.

This way if there is a loop of deadlocks, it will always untangle from one of the threads first. Notice that retry and ordering are both necessary, as the older algorithm was operating under the assumption that at the point of merge we do need to locks on both branches.

If we do not require both branches to be snapshots, then  we may be attaching the head to an older value that sits below the current branch, thus missing some merges. 

It is the lack of retry in your algorithm makes suspect there are corner cases. How do you guarantee if one of the branch advances the merge is still 'correct'? I think the ordering flip is thekey, but the ordering re-arrengement is getting into my way of understanding your algorithm.